src/groth16/aggregate/commit.rs

```diff
@@ -0,0 +1,153 @@
+/// TODO global commitment doc and usage
+use ff::{Field, PrimeField};
+use groupy::{CurveAffine, CurveProjective};
+use rayon::prelude::*;
+
+use crate::bls::Engine;
+use crate::groth16::aggregate::inner_product;
+
+/// Key is a generic commitment key that is instanciated with g and h as basis,
+/// and a and b as powers.
+#[derive(Clone, Debug)]
+pub struct Key<G: CurveAffine> {
+    /// Exponent is a
+    pub a: Vec<G>,
+    /// Exponent is b
+    pub b: Vec<G>,
+}
+/// Commitment key used by the "single" commitment on G1 values as
+/// well as in the "pair" commtitment.
+/// It contains $\{h^a^i\}_{i=1}^n$ and $\{h^b^i\}_{i=1}^n$
+pub type VKey<E> = Key<<E as Engine>::G2Affine>;
+
+/// Commitment key used by the "pair" commitment. Note the sequence of
+/// powers starts at $n$ already.
+/// It contains $\{g^{a^{n+i}}\}_{i=1}^n$ and $\{g^{b^{n+i}}\}_{i=1}^n$
+pub type WKey<E> = Key<<E as Engine>::G1Affine>;
+
+impl<G> Key<G>
+where
+    G: CurveAffine,
+{
+    /// Returns true if commitment keys have the exact required length.
+    /// It is necessary for the IPP scheme to work that commitment
+    /// key have the exact same number of arguments as the number of proofs to
+    /// aggregate.
+    pub fn has_correct_len(&self, n: usize) -> bool {
+        self.a.len() == n && self.b.len() == n
+    }
+
+    /// scale returns both vectors scaled by the given vector entrywise.
+    /// In other words, it returns $\{v_i^{s_i}\}$
+    pub fn scale(&self, s_vec: &[G::Scalar]) -> Self {
+        let (a, b) = self
+            .a
+            .par_iter()
+            .zip(self.b.par_iter())
+            .zip(s_vec.par_iter())
+            .map(|((ap, bp), si)| {
+                let v1s = mul!(ap.into_projective(), si.into_repr()).into_affine();
+                let v2s = mul!(bp.into_projective(), si.into_repr()).into_affine();
+                (v1s, v2s)
+            })
+            .unzip();
+
+        Self { a: a, b: b }
+    }
+
+    /// split returns the left and right commitment key part. It makes copy.
+    /// TODO: remove the copy
+    pub fn split(&self, at: usize) -> (Self, Self) {
```

Comment:
This will do the trick of consuming and splitting 

```rust
fn split(mut source: Vec<usize>, at: usize) -> (Vec<usize>, Vec<usize>) {
    let b = source.split_off(at);
    (source, b)
}
```

Reply:
Ahh perfect thank you :)